Added support for changing uid/gid (fix #157)

stephanvl opened the pull request:
This PR adds support for changing the uid/gid of a file/directory.

Changing the uid/gid of a file is only possible as the root user and will fail silently if this run as a non root user.

What do you think we should do in this case?

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
 }
 
+function getOwnerDiff(fsStat, vinylStat) {
+  if (/^win/.test(process.platform)) {
```
there should be no reason to do this.  In `updateMetadata`, we exit early on windows (it doesn't have `process.getuid()`)

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+    return; // Always return undefined on windows.
+  }
+
+  if (typeof vinylStat.uid !== 'number' &&
```
this should probably be abstracted to a method called `isValidUid()` incase we need to add extra logic and it keeps it consistent with `getTimesDiff`

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+  }
+
+  if (typeof vinylStat.uid !== 'number' &&
+      typeof vinylStat.gid !== 'number') {
```
this should probably be abstracted to a method called `isValidGid()` incase we need to add extra logic and it keeps it consistent with `getTimesDiff`

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+
+  var uid = fsStat.uid; // Default to current uid.
+  if (typeof vinylStat.uid === 'number' &&
+      vinylStat.uid >= 0) {
```
this can go in the `isValidUid` method

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+
+  var gid = fsStat.gid; // Default to current gid.
+  if (typeof vinylStat.gid === 'number' &&
+      vinylStat.gid >= 0) {
```
this can go in the `isValidGid` method.  If these are exactly the same, maybe there's a better name to give them and use it for both.

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+  }
+
+  if (uid < 0) {
+    uid = undefined;
```
this should be defaulted when `isValid...` is checked

phated commented on lib/file-operations.js:
```diff
@@ -66,6 +66,48 @@ function getTimesDiff(fsStat, vinylStat) {
+    uid = undefined;
+  }
+  if (gid < 0) {
+    gid = undefined;
```
this should be defaulted when `isValid...` is checked

phated commented on lib/file-operations.js:
```diff
@@ -149,9 +200,24 @@ function updateMetadata(fd, file, callback) {
           file.stat.mtime = timesDiff.mtime;
         }
+        if (ownerDiff) {
+          owner(fchmodErr, futimesErr);
```
this should probably be `||` like the other callback but i'm not sure

phated replied:
missing the `return` here so the `callback` below would always be called at the wrong time

phated commented on test/file-operations.js:
```diff
@@ -333,6 +334,221 @@ describe('getTimesDiff', function() {
+
+    var result = getOwnerDiff(fsStat, vfsStat);
+
+    expect(result).toEqual(expected);
```
this is failing on appveyor.

phated commented on test/file-operations.js:
```diff
@@ -333,6 +334,221 @@ describe('getTimesDiff', function() {
+      uid: 1001,
+      gid: 1000,
+    };
+    if (isWindows) {
```
If Windows doesn't provide a `uid` or `gid` when `fs.stat` is called, we might as well skip all these tests.  I will have to check this later.

phated commented:
@stephanvl very nice work.  I just had some suggestions and the appveyor CI is still failing.  Let me know once you've had a chance to clean it up.

stephanvl commented:
@phated I removed the latter 2 commits which added the windows specific code, added a `isValidUnixId` method to check the validity of the uid and gid and I added the return statement to fix the callback issue.

I also updated the `owner` method to expect one error instead of separate errors for the owner and times.

phated commented on test/file-operations.js:
```diff
@@ -333,6 +362,206 @@ describe('getTimesDiff', function() {
+    };
+    var expected = {
+      uid: 1001,
+      gid: undefined,
```
fs.fchown throws if uid or gid is undefined.  Do we need to handle this by returning `undefined` if we don't have a complete diff?

phated commented:
@stephanvl this looks near perfect.  Just 1 thing that I noticed while poking around in the node repl.  Commented inline.

stephanvl commented:
@phated Good catch, I will change it to return undefined when either uid or gid is undefined as there is no way of knowing what it should be at that point.

Then there is only the question what we should do about the fact that chown only works when the user is root (fail silently, throw an error or log a message and continue)?

phated commented:
@stephanvl we already handle that with our `isOwner` check inside the `updateMetadata` function.  It checks the user owns the file or is effectively root.

stephanvl commented:
@phated but you must be root for chown to work, being the owner of a file is not enough.

https://en.wikipedia.org/wiki/Chown

phated commented:
Those errors should be bubbled up, I believe.

stephanvl commented:
graceful-fs gracefully ignores EPERM errors, which changing the owner as a non root user is

phated commented:
👍  I'm fine with that then

phated commented:
@stephanvl amazing work!  Thanks a ton
